src/backend/gl/src/device.rs

```diff
@@ -1096,38 +1121,56 @@ impl d::Device<B> for Device {
 
     unsafe fn unmap_memory(&self, memory: &n::Memory) {
         let gl = &self.share.context;
-        let buffer = match memory.first_bound_buffer.get() {
-            None => panic!("No buffer has been bound yet, can't map memory!"),
-            Some(other) => other,
-        };
-        let target = glow::PIXEL_PACK_BUFFER;
+        let buffer = memory.buffer.expect("cannot unmap DEVICE_LOCAL memory");
 
-        gl.bind_buffer(target, Some(buffer));
+        gl.bind_buffer(memory.target, Some(buffer));
 
         if self.share.private_caps.emulate_map {
-            let raw = memory.emulate_map_allocation.replace(std::ptr::null_mut());
-            let mapped = slice::from_raw_parts_mut(raw, memory.size as usize);
+            let ptr = memory.emulate_map_allocation.replace(None).unwrap();
+            let allocation = slice::from_raw_parts_mut(ptr, memory.size as usize);
             // TODO: Access
-            gl.buffer_data_u8_slice(target, mapped, glow::DYNAMIC_DRAW);
-            let _ = *Box::from_raw(raw);
+            gl.buffer_data_u8_slice(memory.target, &allocation, glow::DYNAMIC_DRAW);
+            let _ = Box::from_raw(allocation);
         } else {
-            gl.unmap_buffer(target);
+            gl.unmap_buffer(memory.target);
         }
 
-        gl.bind_buffer(target, None);
+        gl.bind_buffer(memory.target, None);
 
         if let Err(err) = self.share.check() {
             panic!("Error unmapping memory: {:?} for memory {:?}", err, memory);
         }
     }
 
-    unsafe fn flush_mapped_memory_ranges<'a, I, R>(&self, _: I) -> Result<(), d::OutOfMemory>
+    unsafe fn flush_mapped_memory_ranges<'a, I, R>(&self, ranges: I) -> Result<(), d::OutOfMemory>
```

Comment:
we are also missing `invalidate_mapped_memory_ranges` implementation, which is required to be sensible for non-coherent memory

Reply:
Added `glInvalidateBufferSubData` to glow and used that (in personal branch), when using emulated memory maps I'm currently doing... nothing?  I guess that's the right thing to do?

Reply:
Well, flush() updates the data from mapping to GPU, while invalidate() should update the data *from* GPU to the mapping. It sounds like this still needs to be done with the emulated maps?

Reply:
Okay yeah that makes sense.  This will also require a new glow method, I'll merge into this PR once that's merged into glow.